crates/bytecode/src/iterator.rs

```diff
@@ -0,0 +1,261 @@
+use crate::{opcode, Bytecode, OpCode};
+
+/// Iterator over opcodes in a bytecode, skipping immediates.
+///
+/// This allows you to iterate through the actual opcodes in the bytecode,
+/// without dealing with the immediate values that follow PUSH instructions.
+#[derive(Debug, Clone)]
+pub struct BytecodeIterator<'a> {
+    /// Reference to the underlying bytecode bytes
+    bytes: &'a [u8],
+    /// Current position in the bytecode
+    position: usize,
+    /// End position in the bytecode (to handle original length for legacy bytecode)
+    end: usize,
+}
+
+impl<'a> BytecodeIterator<'a> {
+    /// Creates a new iterator from a bytecode reference.
+    pub fn new(bytecode: &'a Bytecode) -> Self {
+        let bytes = bytecode.bytecode();
+        let end = match bytecode {
+            Bytecode::LegacyAnalyzed(analyzed) => analyzed.original_len(),
+            Bytecode::Eip7702(_) => 0,
+            _ => bytes.len(),
+        };
+
+        Self {
+            bytes: bytes.as_ref(),
+            position: 0,
+            end,
+        }
+    }
+
+    /// Returns the current position in the bytecode.
+    pub fn position(&self) -> usize {
+        self.position
+    }
+
+    /// Skips to the next opcode, taking into account PUSH instructions.
+    pub fn skip_to_next_opcode(&mut self) {
+        if self.position >= self.end {
+            return;
+        }
+
+        let opcode = self.bytes[self.position];
+        self.position += 1;
+
+        if opcode::OpCode::new(opcode).is_none() {
+            // Unknown opcode, return in that case
+            return;
+        }
+
+        // Special case: RJUMPV has a variable number of immediates
+        if opcode == opcode::RJUMPV {
+            if self.position < self.end {
+                // The first immediate byte is the max_index
+                let max_index = self.bytes[self.position] as usize;
+                let rjumpv_additional_immediates = (max_index + 1) * 2; // Including the max_index byte itself
+
+                // Skip the immediate bytes
+                self.position =
+                    core::cmp::min(self.position + 1 + rjumpv_additional_immediates, self.end);
+            }
+        } else {
+            // For all other opcodes, use the immediate_size from OPCODE_INFO
+            if let Some(opcode_info) = opcode::OPCODE_INFO[opcode as usize] {
+                let immediate_size = opcode_info.immediate_size() as usize;
+                if immediate_size > 0 {
+                    self.position = core::cmp::min(self.position + immediate_size, self.end);
+                }
+            }
+        }
```

Comment:
We can refactor this in more redable form

```suggestion

        // Special case: RJUMPV has a variable number of immediates
        if opcode == opcode::RJUMPV {
            if let(max_index) = self.bytes.get(self.position) {
                 self.position += max_index * 2; // max_index byte is included later with OPCODE_INFO
            }
        }
        // skip opcode immediated
        self.position += opcode::OPCODE_INFO[opcode as usize].map(|info| info.immediate_size()).unwrap_or_default();
```

Reply:
yeah thats better